### PR TITLE
[merged] Add support for --unshare-cgroup

### PR DESCRIPTION
**Pull request by mrunalp**
This requires linux kernel version 4.6 or higher

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
We'll probably need to conditionally compile this in. Should I follow the model of HAVE_SELINUX by testing for the presence of `/proc/self/ns/cgroup` ?

**Comment by alexlarsson** [comment]
Well, you clearly can't just use CLONE_NEWCGROUP like that and expect it to build on older systems. However, its also not realistic to believe that if it was found in the headers at build-time then its going to be there in the kernel at runtime.

I think we should keep a local version of CLONE_NEWCGROUP:
```
#ifndef CLONE_NEWCGROUP
#define CLONE_NEWCGROUP		0x02000000
#endif
```
And then check at runtime for `/proc/self/ns/cgroup` to see if --unshare-cgroup is supported.

The question is what to do if its not supported though. We could just die with an error, but it seems in some situations you may want to unshare cgroups if possible, and otherwise ignore that. That way you can hide other cgroups from the sandbox. This is pretty similar to the "magic" default for --unshare-user we do in the non-privileged case. Maybe we should standardize how such optional things look in the arguments.

--try-unshare-cgroup ?

**Comment by mrunalp** [comment]
I have updated the PR to add the definition for CLONE_NEWCGROUP and check for /proc/self/ns/cgroup. 
I thinking the --try-feature depends on how portable we want the commands to be. 
Failing if given a non-supported flag is also an option.

File: bubblewrap.c [hunk 1]
```diff
@@ -1313,6 +1323,20 @@ main (int argc,
...
+      {
+        if (errno == ENOENT)
+          {
+            die ("Cannot create new cgroup namespace because the kernel does not support it");
+          }
+        else
+          {
+            die_with_error ("stat on /proc/self/ns/cgroup failed");
+          }
+      }
```
Comment (alexlarsson) [inline bubblewrap.c]: Coding style is to not have brackets around single-line if/else branches (unless one of the branches need it, then both should have it).

File: bubblewrap.c [hunk 2]
```diff
@@ -1313,6 +1323,20 @@ main (int argc,
     clone_flags |= CLONE_NEWIPC;
   if (opt_unshare_uts)
     clone_flags |= CLONE_NEWUTS;
+  if (opt_unshare_cgroup) {
```
Comment (alexlarsson) [inline bubblewrap.c]: brackets on new line

**Comment by alexlarsson** [comment]
> I thinking the --try-feature depends on how portable we want the commands to be. 
> Failing if given a non-supported flag is also an option. 

I think we clearly want to have a way to say "i really need cgroup namespaces, fail if not". That is going to be required if your app is going to set up cgroup stuff. However, I can also see cases where it makes sense to enable cgroup namespaces if possible, to increase isolation (can't see other cgroups), but where the lack of this is not a problem. 

In such a setup you'd have to try first with --unshare-cgroup and if that fails try without it. This is pretty icky, but also doesn't scale to the case where multiple argument are optional in this way. So, as this is likely to come up for other features, maybe we should standardize on how that would look in the command line arguments.

**Comment by mrunalp** [comment]
I agree that we should have a way to optionally apply a feature if possible. The question is should we support both flags then?

**Comment by mrunalp** [comment]
I have added a commit to support `try-unshare-cgroup`

**Comment by rhatdan** [comment]
Can we change this to unshare-cgroup-try unshare-cgroup-optional or something like that,  Want to keep options alphabetically together, when possible.  This would make command completion and man pages easier to read and discover options.

**Comment by alexlarsson** [comment]
Yeah, of those prefer --unshare-cgroup-try, although I'm hoping someone can come up with a better name.

**Comment by mrunalp** [comment]
Updated to rename the option to `--unshare-cgroup-try`. I can't think of any suitable alternatives :/

**Comment by mrunalp** [comment]
Rebased.

**Comment by alexlarsson** [comment]
@cgwalters-bot r+  		d9a5603

**Pull request closed without merging**